File: models/scalar_ou_lqg/_dynamics.py
"""Pure-JAX dynamics for the scalar OU LQG model.

State: y = [x]                     (1-D)
Drift: dx/dt = -a x + b u(t)
Diffusion: sigma_w
Obs: y_k = x_k + N(0, sigma_v^2)

Discrete-time Euler-Maruyama:
    x_{k+1} = (1 - a*dt) x_k + b*dt*u_k + sqrt(dt)*sigma_w*xi_k

The control input u(t) is a deterministic exogenous schedule; for the
filter side we set u_k = 0 (open-loop simulation). When the control
side replaces u_k, it's read from grid_obs['u_value'][k].
"""

from __future__ import annotations

import jax
import jax.numpy as jnp

from smc2fc._likelihood_constants import HALF_LOG_2PI


# Parameter index map. Set in estimation.py and passed in via pi.

def drift_jax(y, t, params, frozen, aux, pi):
    """Scalar OU drift: dx/dt = -a*x + b*u."""
    del frozen, t
    x = y[0]
    a = params[pi['a']]
    b = params[pi['b']]
    u_t = aux['u_at_t']
    return jnp.array([-a * x + b * u_t])


def diffusion(params, frozen, pi):
    """Diagonal diffusion sqrt(noise temperature) — single state, sigma_w."""
    del frozen
    return jnp.array([params[pi['sigma_w']]])


def imex_step_deterministic(y, t, dt, params, frozen, aux, pi):
    """Explicit Euler step on the drift."""
    return y + dt * drift_jax(y, t, params, frozen, aux, pi)


def imex_step_stochastic(y, t, dt, params, sigma_diag, noise, frozen, aux, pi):
    """Euler-Maruyama step. Returns (y_next, mu_prior, var_prior)."""
    y_det = imex_step_deterministic(y, t, dt, params, frozen, aux, pi)
    y_next = y_det + sigma_diag * jnp.sqrt(dt) * noise
    mu_prior = y_det
    var_prior = (sigma_diag ** 2) * dt
    return y_next, mu_prior, var_prior


def obs_log_prob(y, grid_obs, k, params, pi):
    """Gaussian log-pdf for scalar observation at step k."""
    sigma_v = params[pi['sigma_v']]
    x_pred = y[0]
    resid = grid_obs['obs_value'][k] - x_pred
    return grid_obs['obs_present'][k] * (
        -0.5 * (resid / sigma_v) ** 2 - jnp.log(sigma_v) - HALF_LOG_2PI
    )

File: version_1_Julia/models/scalar_ou_lqg/_dynamics.jl
# Pure-Julia dynamics for the scalar OU LQG model.
# Mirrors version_1/models/scalar_ou_lqg/_dynamics.py line-by-line.
#
# State:     y = [x]                     (1-D)
# Drift:     dx/dt = -a x + b u(t)
# Diffusion: σ_w
# Obs:       y_k = x_k + N(0, σ_v²)
#
# Discrete-time Euler-Maruyama:
#     x_{k+1} = (1 - a·dt) x_k + b·dt·u_k + √dt·σ_w·ξ_k

module Dynamics

const HALF_LOG_2PI = 0.9189385332046727   # 0.5 * log(2π)

"""
    drift(y, t, params, u_t)

Scalar OU drift `dx/dt = -a·x + b·u`. `params` indexes `:a, :b`.
"""
@inline function drift(y::AbstractVector, t::Real, params, u_t::Real)
    return [-params[:a] * y[1] + params[:b] * u_t]
end

"""
    diffusion_diagonal(params)

Diagonal diffusion `[σ_w]` (single state).
"""
@inline diffusion_diagonal(params) = [params[:sigma_w]]

"""
    em_step(y, t, dt, params, u_t, ξ)

Explicit Euler-Maruyama step. Returns the new state vector.
"""
@inline function em_step(y::AbstractVector, t::Real, dt::Real, params,
                          u_t::Real, ξ::Real)
    drift_y = drift(y, t, params, u_t)
    σ       = diffusion_diagonal(params)
    return y .+ dt .* drift_y .+ σ .* sqrt(dt) .* ξ
end

"""
    obs_log_prob(y, y_obs::Real, σ_v::Real)

Gaussian log-likelihood of a scalar observation given state `y`.
"""
@inline function obs_log_prob(y::AbstractVector, y_obs::Real, σ_v::Real)
    resid = y_obs - y[1]
    return -0.5 * (resid / σ_v)^2 - log(σ_v) - HALF_LOG_2PI
end

end # module Dynamics
